ledger/src/blockstore_db.rs

```diff
@@ -99,6 +111,222 @@ const PROGRAM_COSTS_CF: &str = "program_costs";
+    pub block_cache_capacity: i64,
+    // The memory size used by the column family in the block cache.
+    // RocksDB's internal property key: "rocksdb.block-cache-usage"
+    pub block_cache_usage: i64,
+    // The memory size used by the column family in the block cache where
+    // entries are pinned.
+    // RocksDB's internal property key: "rocksdb.block-cache-pinned-usage"
+    pub block_cache_pinned_usage: i64,
+
+    // The estimated memory size used for reading SST tables in this column
+    // family such as filters and index blocks. Note that this number does not
+    // include the memory used in block cache.
+    // RocksDB's internal property key: "rocksdb.estimate-table-readers-mem"
+    pub estimate_table_readers_mem: i64,
+
+    // Flush and compaction
+
+    // A 1 or 0 flag indicating whether a memtable flush is pending.
+    // If this number is 1, it means a memtable is waiting for being flushed,
+    // but there might be too many L0 files that prevents it from being flushed.
+    // RocksDB's internal property key: "rocksdb.mem-table-flush-pending"
+    pub mem_table_flush_pending: i64,
+
+    // A 1 or 0 flag indicating whether a compaction job is pending.
+    // If this number is 1, it means some part of the column family requires
+    // compaction in order to maintain shape of LSM tree, but the compaction
+    // is pending because the desired compaction job is either waiting for
+    // other dependnent compactions to be finished or waiting for an available
+    // compaction thread.
+    // RocksDB's internal property key: "rocksdb.compaction-pending"
+    pub compaction_pending: i64,
+
+    // The number of compactions that are currently running for the column family.
+    // RocksDB's internal property key: "rocksdb.num-running-compactions"
+    pub num_running_compactions: i64,
+
+    // The number of flushes that are currently running for the column family.
+    // RocksDB's internal property key: "rocksdb.num-running-flushes"
+    pub num_running_flushes: i64,
+
+    // FIFO Compaction related
+
+    // returns an estimation of the oldest key timestamp in the DB. Only vailable
+    // for FIFO compaction with compaction_options_fifo.allow_compaction = false.
+    // RocksDB's internal property key: "rocksdb.estimate-oldest-key-time"
+    pub estimate_oldest_key_time: i64,
+
+    // Misc
+
+    // The accumulated number of RocksDB background errors.
+    // RocksDB's internal property key: "rocksdb.background-errors"
+    pub background_errors: i64,
+}
+
+impl BlockstoreRocksDbColumnFamilyMetrics {
+    /// Report metrics with the specified metric name and column family tag.
+    /// The metric name and the column family tag is embeded in the parameter
+    /// `metric_name_and_cf_tag` with the following format.
+    ///
+    /// For example, "blockstore_rocksdb_cfs,cf_name=shred_data".
+    pub fn report_metrics(&self, metric_name_and_cf_tag: &'static str) {
+        datapoint_info!(
```

Comment:
Metrics look good, I would just ensure that these aren't being reported too often and overwhelm influx. Reporting an aggregate every few seconds should be ok.

Reply:
Yep, currently, the reporting period is every one second for the column family properties.  The read-side performance metric is for once for every 100 reads.

Reply:
Every 10 seconds no?
https://github.com/solana-labs/solana/blob/c8c3c4359f5bbe698398d4eea36a24ba986aeee5/core/src/ledger_metric_report_service.rs#L16-L17

Reply:
This one is for the column family properties.    Let me rename the const a bit to avoid confusion.

I feel the reads should be sampling-based, but we definitely want to make sure we don't overwhelm the load.

But looks like we are doing unsampled [blockstore-rpc-call](https://github.com/solana-labs/solana/blob/master/ledger/src/blockstore.rs#L1899-L1907) datapoints here in several function calls, and we are [reporting around 15k per second](https://metrics.solana.com:8888/sources/1/chronograf/data-explorer?query=SELECT%20count%28%22method%22%29%20AS%20%22method-count%22%20FROM%20%22mainnet-beta%22.%22autogen%22.%22blockstore-rpc-api%22%20WHERE%20time%20%3E%20%3AdashboardTime%3A%20AND%20time%20%3C%20%3AupperDashboardTime%3A%20GROUP%20BY%20time%28%3Ainterval%3A%29%20FILL%28null%29). so I guess we 1/100 sample here is probably okay.

Btw, I feel those blockstore-rpc-api should definitely measure the get time and offer group-by the type of rpc call.

Reply:
@yhchiang-sol What kind of testings do you do before merging these commits to ensure that these changes are not adding significant overheads to hot paths?